framework/src/convergence/ResidualConvergence.C

```diff
@@ -0,0 +1,371 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+// MOOSE includes
+#include "ResidualConvergence.h"
+#include "FEProblemBase.h"
+#include "PetscSupport.h"
+#include "Executioner.h"
+#include "PerfGraphInterface.h"
+#include "NonlinearSystemBase.h"
+#include "ActionWarehouse.h"
+
+// PETSc includes
+#include <petscsnes.h>
+#include <petscksp.h>
+#include <petscdm.h>
+
+// PetscDMMoose include
+#include "PetscDMMoose.h"
+
+registerMooseObject("MooseApp", ResidualConvergence);
+
+InputParameters
+ResidualConvergence::validParams()
+{
+  InputParameters params = Convergence::validParams();
+  params += FEProblemSolve::residualConvergenceParams();
+
+  params.addClassDescription(
+      "Checks convergence based on absolute and relative error of the residual.");
+
+  return params;
+}
+
+ResidualConvergence::ResidualConvergence(const InputParameters & parameters)
+  : Convergence(parameters),
+    _fe_problem(*getCheckedPointerParam<FEProblemBase *>("_fe_problem_base"))
+{
+  EquationSystems & es = _fe_problem.es();
+
+  if (isParamSetByUser("l_tol"))
+  {
+    es.parameters.set<Real>("linear solver tolerance") = getParam<Real>("l_tol");
+    _l_tol = getParam<Real>("l_tol");
+  }
+  else
+  {
+    _l_tol = es.parameters.get<Real>("linear solver tolerance");
+  }
```

Comment:
```suggestion
    _l_tol = es.parameters.get<Real>("linear solver tolerance");
```
same for all the others pls